server/src/main/java/org/elasticsearch/cluster/ClusterModule.java

```diff
@@ -206,6 +207,7 @@ public IndexNameExpressionResolver getIndexNameExpressionResolver() {
         addAllocationDecider(deciders, new ThrottlingAllocationDecider(settings, clusterSettings));
         addAllocationDecider(deciders, new ShardsLimitAllocationDecider(settings, clusterSettings));
         addAllocationDecider(deciders, new AwarenessAllocationDecider(settings, clusterSettings));
+        addAllocationDecider(deciders, new NodeShutdownAllocationDecider());
```

Comment:
These deciders are actually loosely in order of performance, because regular decision making short-circuits to avoid doing the processing if it's not needed.

So I think should should be moved up to right after the `RestoreInProgressAllocationDecider`, because it's fairly simple and no need to calculate filtering rules if the node can't have any data because it's being shut down anyway.